[telegram] Telegram Binding initial contribution

ZzetT opened the pull request:
This binding shall work as a replacement for the old OH1 Telegram Actions.
It supports not only sending messages, but also receiving them.  The base for the implementation is the [TelegramBot library](https://github.com/rubenlagus/TelegramBots).

User Belgadon (Jens) initially started the work on that binding which still used the old OH1 telegram actions for sending messages. I removed this dependency by using the new OH2 actions and "tried" to migrate it to the new build environement.

The development of that binding was discussed in that [thread](https://community.openhab.org/t/new-telegram-binding-and-updated-action-tester-and-feedback-welcome).

Although I still didn't completly manage to get it running with the latest build environemnt, I was told in the forum to create a PR anyway.
The problems that I still have are either OSGI errors when I click "Resolve" in the bnd demo app or 
when I add:
```
  <properties>
    <bnd.importpackage>org.telegram.telegrambots.*;resolution:="optional"</bnd.importpackage>
  </properties>  
```

then I get ClassNotFoundExceptions at runtime after adding a new Telegram thing.
One more thing to note here, IMHO it should not be necessary to also add a dependency to "telegrambots-meta" in the pom.xml (only telegrambots should be required), but if I remove that I get an error that this is missing.

Please note that the binding is using an apache client instead of a jetty client, but this issue was discussed already and agreed by @davidgraeff .

So, before continuing the work on more features, we need the approval that the library can be integrated into the OH2 environment and what needs to be configured to get it running again. Then, I hope that we can add a few more features in separate PRs for this binding.
Especially two things are currently missing compared to the old Telegram actions:

- sending photos located on a webserver that requires username/password
- support of html markup (should be easy to realize)

Currently, a beta version of that binding was released to several testers in the forum which however was only based on the old build environement.

Signed-off-by: Alexander Krasnogolowy <alexkrasno@hotmail.com>

J-N-K commented on bundles/org.openhab.binding.telegram/NOTICE:
```diff
@@ -0,0 +1,13 @@
+
+== Source Code
+
+https://github.com/openhab/openhab2-addons
```
please add the 3rdparty-library here

ZzetT replied:
ok, I added this information

J-N-K commented on bom/openhab-addons/pom.xml:
```diff
@@ -1,965 +1,970 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Please check why this file (and CODEOWNERS) show up as fully changed.

ZzetT replied:
fixed, was a problem with the line endings

ZzetT commented:
> The library is MIT licensed, I do not see any problem in using it. I didn't check why you use apache http client, can you add a note to this PR? It's easier to find in case someone wants to look it up.

I'm not using the apache http client, but the Telegram library does. I tried to rewrite it, but there were a lot of changes necessary and it would be a big effort to keep it up to date with the original library.

J-N-K commented:
Ok with me as well. Don‘t rewrite external libs.

ZzetT commented:
> Adding optional to bnd.importpackage is wrong. You need these classes, they are not optional but required. As a workaround you can add the libraries to the /lib directory. But please do not commit these files.

But then do I need this bnd importpackage at all? I found only very few bindings that have this. And some others also use an external library but don't have this bnd statement.

Sorry, I accidentally removed it now. Shall I just remove the optional attribute?

J-N-K commented:
This is only for very special cases with embedded libraries. You don‘t need it.

J-N-K commented:
Is there anything missing from me here?

kaikreuzer commented on bundles/org.openhab.extensionservice.marketplace.automation/pom.xml:
```diff
@@ -11,7 +11,7 @@
   <artifactId>org.openhab.extensionservice.marketplace.automation</artifactId>
 
-  <name>openHAB Add-ons :: Extension Service :: Eclipse IoT Marketplace (Automation)</name>
+  <name>openHAB Add-ons :: Bundles :: Extension Service :: Eclipse IoT Marketplace (Automation)</name>
```
Could you please remove all the changes in pom.xml of other bundles from this PR?

ZzetT replied:
That's weird. These changes are actually not from me, but are from the last commit (a248d5d30d5a0230eefaa42bd5e7544e212471ba) *before* I forked the repo. However, I saw this automatic error message here ""Signed-off-by" needed on" and then I executed a command to add the signed-off statement to all my previous commits on the forked repo. It seems that unfortunately, the signed-off was also added to the last commit from the openhab repo which was not from me. Not sure if the changes will go away if I just revert the comment.

kaikreuzer replied:
Could you try to rebase the branch then? Btw, I see that you created the PR from your master branch - please better always work on a dedicated feature branch, that will make your life easier, if you ever work on multiple PRs at the same time.

ZzetT replied:
Yes, I agree, I'll use a branch instead of a master next time. I was now able to remove my signed-off  from commit 8d246199736567d4a3ab938eeae40d71e1d1214d. However, I now see all commits in this PR twice. I'm not super experienced with git and it's also my first PR and don't want to make it worse. I guess I can't rebase because I'm already on the master? The only clean option I could think of is to re-fork openhab2-addons again, create a branch and merge my changes to that new branch. However, I think I cannot change the source of this PR which is currently the master branch. Sorry for the muddle.

ZzetT replied:
I did a rebase so the changes are gone now.

kaikreuzer commented:
> Is there anything missing from me here?

@J-N-K There is still a "changes requested" review from you here - so yes, if you are satisfied, you should change it to an approval.

J-N-K commented:
Mhm. Is there any better way to remove "requests changes"? I only looked at the infrastructure part, not at the code.

kaikreuzer commented:
There is usually also a "dismiss review" button as an option.

ZzetT commented:
> Adding `optional` to `bnd.importpackage` is wrong. You need these classes, they are not `optional` but required. As a workaround you can add the libraries to the `/lib` directory. But please do not commit these files.

Hi @J-N-K , I downloaded the jar with depedencies from maven cnetral [(telegrambots-4.2-jar-with-dependencies.jar](http://central.maven.org/maven2/org/telegram/telegrambots/4.2/telegrambots-4.2-jar-with-dependencies.jar)) and put that into the lib folder of the binding and removed the dependencies from features.xml as well as from the pom.xml (only locally). But even then I get osgi errors:

```
[ERROR] Failed to execute goal org.apache.karaf.tooling:karaf-maven-plugin:4.2.4:verify (karaf-feature-verification) on project org.openhab.binding.telegram: Feature resolution failed for [openhab-binding-telegram/2.5.0.SNAPSHOT]
[ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-telegram; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-telegram)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-telegram/2.5.0.SNAPSHOT: missing requirement [openhab-binding-telegram/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.telegram; type=osgi.bundle; version="[2.5.0.201906091517,2.5.0.201906091517]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.telegram/2.5.0.201906091517: missing requirement [org.openhab.binding.telegram/2.5.0.201906091517] osgi.wiring.package; filter:="(osgi.wiring.package=afu.com.sun.source.tree)"]]
```

I still don't understand it, because when I just put the mentioned jar in a simple eclipse project (without OH)  and send a message it works like a charme. 
What do you recommend to proceed here? I just can't get it running with the current build environment.

J-N-K commented:
I‘ll checkout a copy and have a look.. I can‘t promise to get it ready today.

ZzetT commented:
> I‘ll checkout a copy and have a look.. I can‘t promise to get it ready today.

Thanks, that would be great! No hurry!

J-N-K commented:
After downloading what is committed here: What exactly is the problem? That you can't resolve it in the IDE? 

Download `.jar`, not the dependencies, download `telegrambots` and `telegrambots-meta`. Remove the feature-dependencies and the pom-dependencies. Then you can compile and feature resolution works.

BUT: this is in no way possible to merge, this is only a workaround for debugging in the IDE.

ZzetT commented:
> After downloading what is committed here: What exactly is the problem? That you can't resolve it in the IDE?

I can't event compile it with maven.

> Download `.jar`, not the dependencies, download `telegrambots` and `telegrambots-meta`. Remove the feature-dependencies and the pom-dependencies. Then you can compile and feature resolution works.
 
Ok, I did that and downloaded [telegrambots-4.2.jar](http://central.maven.org/maven2/org/telegram/telegrambots/4.2/telegrambots-4.2.jar) and [telegrambots-meta-4.2.jar](http://central.maven.org/maven2/org/telegram/telegrambots-meta/4.2/telegrambots-meta-4.2.jar). But then I get another osgi.wiring.package error for com.fasterxml.jackson.annotations. I tried it with a fresh install of OH2 development environment. Do you have any additional configuration?

I tried to solve it myself by adding com.fasterxml.jackson.core to the feature.xml as well as to the pom.xml. It solved that problem but then it came it with another missing dependency. 
My pom.xml looks like this now:
```
  <dependencies>
    <dependency>
      <groupId>org.glassfish</groupId>
      <artifactId>javax.json</artifactId>
      <version>1.0.4</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-core</artifactId>
      <version>2.9.8</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-annotations</artifactId>
      <version>2.9.0</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <version>2.9.8</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.glassfish.jersey.containers</groupId>
      <artifactId>jersey-container-grizzly2-http</artifactId>
      <version>2.25.1</version>
      <scope>provided</scope>
    </dependency>    
    <dependency>
        <groupId>org.glassfish.jersey.core</groupId>
        <artifactId>jersey-server</artifactId>
        <version>2.25.1</version>
    </dependency>
    <dependency>
        <groupId>org.json</groupId>
        <artifactId>json</artifactId>
        <version>20180813</version>
    </dependency>    
  </dependencies>
```

and my features.xml:
 ```
       <feature prerequisite="true">wrap</feature>
        <bundle dependency="true">mvn:org.glassfish/javax.json/1.0.4</bundle>
        <bundle dependency="true">mvn:com.fasterxml.jackson.core/jackson-annotations/2.9.8</bundle>
        <bundle dependency="true">mvn:com.fasterxml.jackson.core/jackson-core/2.9.8</bundle>
        <bundle dependency="true">mvn:com.fasterxml.jackson.core/jackson-databind/2.9.8</bundle>
        <bundle dependency="true">mvn:org.apache.httpcomponents/httpcore-osgi/4.4.9</bundle>
        <bundle dependency="true">mvn:org.apache.httpcomponents/httpclient-osgi/4.5.5</bundle>
```

But now I get this:
`osgi.wiring.package=org.glassfish.grizzly.http.server)(version>=2.3.0)(!(version>=3.0.0)))"]]`

I still don't know how you got it working without all that. 

> BUT: this is in no way possible to merge, this is only a workaround for debugging in the IDE.

Yes, I know, but if I can't get it working myself for debugging puposes, I can't test anything.
So I would be very happy to get it running with the new build environment. 

Of course, if I ever get it running again, my second question would be how do I prove that the things are working on a build server if I can't simulate it in the same way on my local machine? (which is without any lib and dependencies only)

J-N-K commented:
I raised a PR against your branch. I did nothing but cloning your repo and running "mvn cclean install".

ZzetT commented:
Thanks for your support @J-N-K. I agree that with the current branch it is compiling (not sure why not with the current master of OH2 addons, maybe just my fault, maybe stricter checks...).
Anyway, I then added the binding to the pom.xml of the demo app. Otherwise, I'm not able to see it in the bnd run. When I then click "Resolve" I get the error:
```
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=org.openhab.binding.telegram)
          ⇒ [org.openhab.binding.telegram version=2.5.0.201906100917]
              ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.apache.http))
```
I guess I now have to add it to the features, right? ( I tried it but the error is still the same)

J-N-K commented:
Have a look here:

http://central.maven.org/maven2/org/telegram/telegrambots-meta/4.2/telegrambots-meta-4.2.pom
http://central.maven.org/maven2/org/telegram/telegrambots/4.2/telegrambots-4.2.pom

What is listed there (and not `<scope>test</scope>`) is probably needed. Check if an OSGi-version of the bundles is available (in case they are not already OSGi-bundles). If you add dependencies (`<scope>provided</scope>`), it should resolve in bnd. Adding to the `feature.xml` is necessary for the feature-check of KARAF (and using it in OH later) only.

ZzetT commented:
These are the dependencies I ended up with:
```
  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
    <guice.version>4.2.2</guice.version>
    <glassfish.version>2.25.1</glassfish.version>
    <jerseybundle.version>1.19.3</jerseybundle.version>
    <httpcompontents.version>4.5.3</httpcompontents.version>
    <jackson.version>2.9.8</jackson.version>
    <jacksonanotation.version>2.9.0</jacksonanotation.version>
    <json.version>20180813</json.version>
    <guava.version>25.1-jre</guava.version>
    <commonio.version>2.5</commonio.version>
  </properties>
  <dependencies>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-annotations</artifactId>
      <version>${jacksonanotation.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.jaxrs</groupId>
      <artifactId>jackson-jaxrs-json-provider</artifactId>
      <version>${jackson.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <version>${jackson.version}</version>
      <scope>provided</scope>
      <exclusions>
        <exclusion>
          <groupId>com.fasterxml.jackson.core</groupId>
          <artifactId>jackson-annotations</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.glassfish.jersey.media</groupId>
      <artifactId>jersey-media-json-jackson</artifactId>
      <version>${glassfish.version}</version>
      <scope>provided</scope>
      <exclusions>
        <exclusion>
          <groupId>com.fasterxml.jackson.core</groupId>
          <artifactId>jackson-annotations</artifactId>
        </exclusion>
        <exclusion>
          <groupId>com.fasterxml.jackson.jaxrs</groupId>
          <artifactId>jackson-jaxrs-json-provider</artifactId>
        </exclusion>
        <exclusion>
          <groupId>com.fasterxml.jackson.jaxrs</groupId>
          <artifactId>jackson-jaxrs-base</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>org.glassfish.jersey.containers</groupId>
      <artifactId>jersey-container-grizzly2-http</artifactId>
      <version>${glassfish.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.glassfish.jersey.core</groupId>
      <artifactId>jersey-server</artifactId>
      <version>${glassfish.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.json</groupId>
      <artifactId>json</artifactId>
      <version>${json.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.apache.httpcomponents</groupId>
      <artifactId>httpclient</artifactId>
      <version>${httpcompontents.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.apache.httpcomponents</groupId>
      <artifactId>httpmime</artifactId>
      <version>${httpcompontents.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>commons-io</groupId>
      <artifactId>commons-io</artifactId>
      <version>${commonio.version}</version>
      <scope>provided</scope>
    </dependency>

    <dependency>
      <groupId>com.google.inject</groupId>
      <artifactId>guice</artifactId>
      <version>${guice.version}</version>
      <scope>provided</scope>
      <exclusions>
        <exclusion>
          <groupId>com.google.guava</groupId>
          <artifactId>guava</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <version>${jackson.version}</version>
      <scope>provided</scope>
      <exclusions>
        <exclusion>
          <groupId>com.fasterxml.jackson.core</groupId>
          <artifactId>jackson-annotations</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-annotations</artifactId>
      <version>${jacksonanotation.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>com.google.guava</groupId>
      <artifactId>guava</artifactId>
      <version>${guava.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.json</groupId>
      <artifactId>json</artifactId>
      <version>${json.version}</version>
      <scope>provided</scope>
    </dependency>
  </dependencies>
```

and these are my current features:
```
    <feature name="openhab-binding-telegram" description="Telegram Binding" version="${project.version}">
        <feature>openhab-runtime-base</feature>
        <feature prerequisite="true">wrap</feature>
        <bundle dependency="true">mvn:com.fasterxml.jackson.core/jackson-annotations/2.9.8</bundle>
        <bundle dependency="true">mvn:com.fasterxml.jackson.core/jackson-core/2.9.8</bundle>
        <bundle dependency="true">mvn:com.fasterxml.jackson.core/jackson-databind/2.9.8</bundle>
        <bundle dependency="true">wrap:mvn:com.google.inject/guice/4.2.2</bundle>
        <bundle dependency="true">wrap:mvn:com.google.guava/guava/25.1-jre</bundle>
        <bundle dependency="true">mvn:commons-io/commons-io/2.5</bundle>
        <bundle dependency="true">mvn:org.apache.httpcomponents/httpcore-osgi/4.4.6</bundle>
        <bundle dependency="true">mvn:org.apache.httpcomponents/httpclient-osgi/4.5.3</bundle>
        <bundle dependency="true">mvn:org.glassfish.jersey.containers/jersey-container-grizzly2-http/2.25.1</bundle>
        <bundle dependency="true">mvn:org.glassfish.grizzly/grizzly-http-server/2.3.28</bundle>
        <bundle dependency="true">mvn:org.glassfish.hk2.external/javax.inject/2.5.0-b32</bundle>
        <bundle dependency="true">mvn:org.glassfish.jersey.media/jersey-media-json-jackson/2.25.1</bundle>
        <bundle dependency="true">mvn:org.glassfish.jersey.core/jersey-server/2.25.1</bundle>
        <bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.telegram/${project.version}</bundle>
    </feature>
```

I'm adding one feature after the other by checking the transitive compile dependencies provided by mvnrespository.com 

Am I really on the right track? It took quite a long time to grab all those things. Is there not somehow an automated way to create the features?

Btw. I'm trying this again on a newer branch of OH2 addons since the feature verification during maven clean install seems to be better !?

Anyway, I'm getting now the following osgi error:

[ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-telegram; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-telegram)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-telegram/2.5.0.SNAPSHOT: missing requirement [openhab-binding-telegram/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.telegram; type=osgi.bundle; version="[2.5.0.201906112201,2.5.0.201906112201]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.telegram/2.5.0.201906112201: missing requirement [org.openhab.binding.telegram/2.5.0.201906112201] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.glassfish.grizzly.ssl)(version>=2.3.0)(!(version>=3.0.0)))"]]

But for org.glassfish.grizzly.ssl I cannot find further features/dependencies.

Sorry, but I already spent a lot of time to get this running. Is there really not an easier way or a better tutorial? It's already long ago that I developed something related to the actual binding. At the moment I'm not sure if I got completly stuck with the migration.

martinvw commented on bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This looks like a combination of spaces and tabs

martinvw replied:
Tabs are the preferered option here

ZzetT replied:
fixed

martinvw commented on bundles/org.openhab.binding.telegram/pom.xml:
```diff
@@ -0,0 +1,32 @@
+        <version>4.2</version>
+        <scope>provided</scope>
+    </dependency>
+	<dependency>
```
Mix of tabs and spaces, please use spaces in this pom.xml file.

ZzetT replied:
fixed

ZzetT commented:
I didn't give up yet. I think I have now a feature.xml and pom.xml with all dependencies and transitive dependencies.
But still, I get one missing requirement:
`osgi.wiring.package; filter:="(osgi.wiring.package=sun.net)": `
And here I don't know where it comes from. I also couldn't find it in any transitive dependencies.
I put everything on a branch: https://github.com/ZzetT/openhab2-addons/tree/with_lib
It would be great if somebody can check that out and have a look.

A second thing is that the feature.xml looks quite bit. I thought I could replace some of the dependency by adding the "openhab-transport-http" feature, but it seems that the versions don't match. So, I had to explicitly add those necessary feature with the right version.

J-N-K commented:
Which bundle  needs that?

ZzetT commented:
> Which bundle needs that?

If I would know that, I could probably fix it by myself, but the full error message doesn't let you know either:

> [ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-telegram; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-telegram)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-telegram/2.5.0.SNAPSHOT: missing requirement [openhab-binding-telegram/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.telegram; type=osgi.bundle; version="[2.5.0.201906150846,2.5.0.201906150846]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.telegram/2.5.0.201906150846: missing requirement [org.openhab.binding.telegram/2.5.0.201906150846] osgi.wiring.package; filter:="(osgi.wiring.package=sun.net)"]]

I even ran mvn with "-X" option but couldn't find any hints in the logging output.

J-N-K commented:
Ok. `sun.net` is required by the `telegrambots` library if you use a SOCKS proxy. I don't think we use that here, so we can mark that as optional import by adding

```
<bnd.importpackage>sun.net.*;resolution:="optional"</bnd.importpackage>
```

to the properties.

ZzetT commented:
Awesome. That works and now it compiles again with maven. Thanks a lot!

Did you also try to add it to bnd run? There it still fails with:
```
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=org.openhab.binding.telegram)
          ⇒ [org.openhab.binding.telegram version=2.5.0.201906161606]
              ⇒ osgi.wiring.package: (&(osgi.wiring.package=com.google.inject)(&(version>=1.4.0)(!(version>=2.0.0))))
```

ZzetT commented:
Still no news here :(

What I don't understand is what is the difference between the maven build which includes karaf feature verification and runs fine and the bnd run which is also based on karaf but detects other dependency issues?

So, I still get this error:;
`osgi.wiring.package: (&(osgi.wiring.package=com.google.inject)(&(version>=1.4.0)(!(version>=2.0.0))))`
which I don't understand since I have this google inject added to pom.xml and feature.xml. If I add this to the bnd.importpackage as well with 
`com.google.inject.*;resolution:="optional" `

then the one error is gone, but I'll get a new one:
`osgi.wiring.package: (&(osgi.wiring.package=org.apache.commons.io)(&(version>=1.4.0)(!(version>=2.0.0))))`
and so on and so on.

I'm already at the point where I consider to stop the development because it seems that the library is not suitable for using in an OSGi environment. What is your opinion @davidgraeff ? I really would like to finish my work on that binding but I can't find a way out of this dependency hell. And this [feature.xml](https://github.com/ZzetT/openhab2-addons/blob/with_lib/bundles/org.openhab.binding.telegram/src/main/feature/feature.xml) already doesn't look healthy.

davidgraeff commented:
>  bnd run which is also based on karaf 

That's not the case. karaf feature files are only for karaf. To make a pure bnd based start work, you must have correct pom.xml dependencies declared and run the bnd resolver once, to alter the bndrun file in question. It's two independent systems.

I'm also not sure how to proceed with openHAB. Karaf and bndrun doesn't really match.
bnd is implementing what the OSGi committee actually had in mind. They support OSGi repositories, OSGi subsystems and all that jazz for dependency resolution. Karaf predates all of that and their feature files are nowhere else used.

J-N-K commented:
@ZzetT is the latest version committed to your repo? I can try to have a look in the next days.

ZzetT commented:
@J-N-K Yes, the branch [with_lib ](https://github.com/ZzetT/openhab2-addons/tree/with_lib) is up-to-date now. I appreciate your support.

J-N-K commented:
This draws in a lot of dependencies. I do not fully understand why this is all necessary. Is there a special reason why you chose this implementation? I had a quick look at https://github.com/pengrad/java-telegram-bot-api which seems to be a lot less dependency-intensive.

ZzetT commented:
> Is there a special reason why you chose this implementation?

I didn't really chose it. As you can see in that [thread](https://community.openhab.org/t/new-telegram-binding-and-updated-action-tester-and-feedback-welcome) the initial binding was developed by Belgadon. A bit later I took it over/developed it further. With OH2.5 the library was not a problem and a working test version was already released in February. Then, it was said that the apache client of that library was bad. So I rewrote it to jetty. (yes, bad idea). Then, I tried to migrate it to the new build environment and now it seems that the osgi environment of OH3 and the number of dependencies is a showstopper.

Sorry for this long story, but I hope you can imagine my frustration to work on the "infrastructure" for a long time and not on the actual implementation and finally it doesn't work.

Both libraries (the currently used one and the one you poroposed) are also mentioned in https://core.telegram.org/bots/samples. The community for the currently used libary seems to be a bit more active/more contributors, but I can have a look at the other library. But I don't know yet how long a migration will take.

Anyway, it was also important to me to get a confirmation from one of the maintainer that this library is not really suitable for OH.

J-N-K commented:
Wait, you say that the initial version was working fine without any of these dependencies? Which other bindings did you install when using it? Maybe we can exclude even more packages from the osgiified library and reduce the number of needed dependencies.

ZzetT commented:
I just wanted to say that with the old OH environment, it was easier because we just put the telegram JAR with dependencies into it and you got a running binding. With the new environment you have to be very careful what to write in the bnd and pom.xml.

J-N-K commented:
Essentially it's the same. You should not need additional dependencies. If you add all dependencies you put into the lib directory before, it should be all good. If that is not the case, we might Need to exclude some packages from ex/import. 

Can you send me a full list of the bundles you used before the migration? I think that maybe some parts of the telegram bundle are not used by your code but require the most of the dependencies.

ZzetT commented:
[Here](https://github.com/Belgadon/openhab2-addons/tree/8cf2d030d12acf6e987eb08a6df750c4f44a93e7/addons/binding/org.openhab.binding.telegram) you can find the last version of the binding based on the old OH2 environment before migration.

However, even there the [telegrambots-4.1-jar-with-dependencies.jar](https://github.com/Belgadon/openhab2-addons/blob/8cf2d030d12acf6e987eb08a6df750c4f44a93e7/addons/binding/org.openhab.binding.telegram/lib/telegrambots-4.1-jar-with-dependencies.jar) contained "everything" (12MB).

J-N-K commented:
@ZzetT I wouldn't say that it can't be used. Maybe the other lib is a little bit more lightweight. But We have other heavy bindings, too. If you can spend a minute investigating if it is much work to use the other lib, that would be great. But I would not say we can't merge this version.

BTW: the osgi-fied version of the library has been made available. If you rebase your branch on current master, you can delete the embedded library, apply my changes and it should still work in the IDE. Maybe you have to build once with maven before bnd can resolve it for debugging. If there are any further problems, please tell me.

ZzetT commented:
> BTW: the osgi-fied version of the library has been made available.

I still get this error:

> [ERROR] Failed to execute goal on project org.openhab.binding.telegram: Could not resolve dependencies for project org.openhab.addons.bundles:org.openhab.binding.telegram:jar:2.5.0-SNAPSHOT: The following artifacts could not be resolved: org.openhab.osgiify:org.telegram.telegrambots:jar:4.2, org.openhab.osgiify:org.telegram.telegrambots-meta:jar:4.2: Could not find artifact org.openhab.osgiify:org.telegram.telegrambots:jar:4.2 in osgiify (https://dl.bintray.com/openhab/mvn/openhab-osgiify) -> [Help 1]

And I also can't see it in https://bintray.com/openhab/mvn/openhab-osgiify

Do I have to use another repo?

J-N-K commented:
@ZzetT it should work now, Kai has created an automatic build for the osgiify-repo.

ZzetT commented:
Yes, the maven build for the binding works fine again.
But the problem with the bnd run remains. 
Adding it to the demo app and resolving it gives still a lot of "Resolution failed. Capabilities satisfying the following requirements could not be found:" errors:
![bndrun](https://user-images.githubusercontent.com/6074452/61325807-cb4b4c80-a815-11e9-9137-9a5f5389b318.png)

Compiling the demo via mvn shows also a very long list of unresolved requirements 👎

kaikreuzer commented on bundles/org.openhab.binding.telegram/pom.xml:
```diff
@@ -0,0 +1,32 @@
+        <groupId>org.telegram</groupId>
+        <artifactId>telegrambots</artifactId>
+        <version>4.2</version>
+        <scope>provided</scope>
```
You need to remove this scope in order to make it work in the demo app.

kaikreuzer commented on bundles/org.openhab.binding.telegram/pom.xml:
```diff
@@ -0,0 +1,32 @@
+	    <groupId>org.telegram</groupId>
+	    <artifactId>telegrambots-meta</artifactId>
+	    <version>4.2</version>
+	    <scope>provided</scope>
```
You need to remove this scope in order to make it work in the demo app.

ZzetT commented:
@kaikreuzer It still gives me the same errors. After that step I also can't add it to the "Run requirements" anymore because it doesn't appear on the left side in the bnd.run window. I also tried it with the steps that are mentioned in that [post](https://community.openhab.org/t/solved-problem-install-eclipse-ide/75140/5) but also no luck.

By the way, I updated the master branch with the changes from @J-N-K to use the osgi version of the library.

kaikreuzer commented:
@maggu2810 Sorry for the probably silly question: What does one need to do to develop a new binding in the IDE, i.e. make it available as a "run requirement"? If I manually add it there, it just tells me that it cannot be resolved:

![Bildschirmfoto 2019-07-23 um 10 46 14](https://user-images.githubusercontent.com/3244965/61697260-29f15880-ad37-11e9-80e0-b18523894149.png)

maggu2810 commented:
I am not sure if I understand the question fully.

[Here](https://github.com/openhab/openhab2-addons/pull/5677#issuecomment-511965976) the binding is shown in the run requirements there are "just" a few capabilities that cannot be satisfied.

> What does one need to do to develop a new binding in the IDE, i.e. make it available as a "run requirement"?

What have we done to get the Z-Wave binding in the list of available bundles?
We add it to the demo apps dependencies using the scope "runtime".

kaikreuzer commented:
> What have we done to get the Z-Wave binding in the list of available bundles?
> We add it to the demo apps dependencies using the scope "runtime".

ZWave was an exceptional case as it resides in its (bom-less) own repo.
But I found what I was looking for, it is step 3+4 from [these instructions](https://www.openhab.org/docs/developer/ide/eclipse.html#develop-a-new-binding-with-the-eclipse-ide).

kaikreuzer commented:
@ZzetT It seems you have missed my comment https://github.com/openhab/openhab2-addons/pull/5677/files#r304112208 - you have to remove the scope from both dependencies, then they should be successfully resolved.

ZzetT commented:
@kaikreuzer  No, I did that. I just didn't commit those changes, because as far as I understood your comment, this is only something for my local workspace but in the repo the scope should stay to "provided". Did you get it running? When I remove it, I can't add it anymore to "run requirements" in the bnd.run.

ZzetT commented:
Weird, I removed the scope and re-compiled everything. Now I can select it in bnd.run, but then I'm getting other errors… 
![binding](https://user-images.githubusercontent.com/6074452/61789221-6941a700-ae14-11e9-9129-1c9f231f86a1.png)

Is it right to also add the two osgiify libraries to the run requirements?

kaikreuzer commented:
> this is only something for my local workspace but in the repo the scope should stay to "provided".

Pfew, after a hint from @maggu2810, I actually thought that such dependencies must not have the "provided" scope (also not in the repo), but I see that [our documentation says otherwise](https://www.openhab.org/docs/developer/buildsystem.html#adding-dependencies), so I have no clue, what is correct. @maggu2810 Could you help clarifying this question?

> Now I can select it in bnd.run, but then I'm getting other errors…

This looks like something is still outdated. We [saw this a week ago as well](https://github.com/openhab/openhab-distro/issues/927#issuecomment-512363189) on other installations. Not sure how to best resolve this, afair, I got it solved by running `mvn clean install` from within the `launch/app`folder.

> Is it right to also add the two osgiify libraries to the run requirements?

No need, they will be automatically added upon resolve.

> Did you get it running?

I am a step further. When I try to resolve, I am now seeing this:

![Bildschirmfoto 2019-07-24 um 22 19 16](https://user-images.githubusercontent.com/3244965/61826950-ce70b900-ae63-11e9-87e7-6b1f08dd7ab2.png)

Bnd is correct here - the telegram bot libs depend on `org.apache.http`, but such a dependency is not yet declared in the pom of the binding. Not sure about the correct Maven GAV for this (especially which version is required), so I leave this for you to sort out.

maggu2810 commented:
I assume there are several documentation that explain the Maven dependency scope, e.g.

* https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html
* https://www.baeldung.com/maven-dependency-scopes

Here a very short summary from my side -- but I do not guarantee the correctness of this information (the test scope is not contained here to keep it more simple ATM):

compile is the default scope if not explicit set

| scope | transitive | provided on compilation | provided on runtime |
|:---:|:---:|:---:|:---:|
| compile | y | y | y |
| provided | n | y | n |
| runtime | y | n | y |

If B depends on A using a **non** transitive dependency and C depends on B, C does not depend on A and it will not be provided (we assume C does not declare a direct dependency to A itself).

So, for OSGi bundles the usage of `provided` makes a lot of sense, as we would like to be consumed by another one and don't want to force which implementation of an API we depend on must be used by the consumer.

The bnd run Maven plugin uses default all dependencies that are marked as "provided on runtime", so `compile` and `runtime` (this can be changed but IMHO shouldn't if you don't have an explicit reason for).

maggu2810 commented:
It depends on your decision how to align dependencies used by Add-ons.
Would you like that every Add-on is using its separate implementation dependency or should they use some provided ones.
IIRC (I am pretty sure) this has been already comment on another issue / PR (@J-N-K do you remember?). I assume it depends on the "type" of the dependency. For binding specific dependencies you could use compile for something provided by core or a more global scope (stuff used by separate addons and aligned by core or another scope) provided would be the best fit.

> What should we tell people how to deal with the dependencies? The docs say, they should use "provided", but for the demo app we say that we need "compile" (or "runtime") scope.

Tell the people what the Maven scope defines and perhaps elaborate how it affect the demo app.
If only one scope would be correct in all cases I assume the other ones has not been created ever.

J-N-K commented:
While migrating to bnd we decided to make everything „provided“ (except those bundles that are not available on bintray/maven central. Those are still embedded. We should probably rethink that and use „compile“ for those that are only used by one binding (which is probably true for 70% of the dependencies) and only use „provided“ for those that will most likely be reused (e.g. okhttp, jackson, guava, ...)

kaikreuzer commented:
I agree.
Imho, what we had embedded in the past should be "compile" and everything that is shared between bindings is "provided" - in the past those were simply "provided" in the target platform. Now, we would provide those as core features. All provided ones should be available in the demo app by default (just like they were available within the TP in the past).

kaikreuzer commented:
@ZzetT Did you try to [add the Apache HTTP libs as a dependency](https://github.com/openhab/openhab2-addons/pull/5677#issuecomment-514793526)? Should be the last step before we can merge.

ZzetT commented:
Hey @kaikreuzer,

I tried to add this dependency but it didn't change anything. But even if, what about all the other dependencies that we need? @J-N-K  added the osgiify version of the library and removed a lot of those dependencies (see https://github.com/ZzetT/openhab2-addons/commit/1cf6ebc98e49805a7b21e76818e683b7d67b3f4f). So I guess that if we add one, we also have to add all the others again?
@J-N-K was it running for you or did you just try to compile it?

> Should be the last step before we can merge.

I highly doubt that :(  Maybe I'm too pessimistic, but I'm not sure if we are any step further here.

J-N-K commented:
I guess you need to add the bundles in the feature.xml to the pom.xml. I compiled and let the feature-check succeed.

bjoernbrings commented:
@ZzetT Could you succeed with the hint of @J-N-K?

ZzetT commented:
I tried several things to get rid of the org.apache.http warning, but it is still there. Last thing I added:
`    <dependency>
      <groupId>org.apache.httpcomponents</groupId>
      <artifactId>httpclient</artifactId>
      <version>4.5.5</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.apache.httpcomponents</groupId>
      <artifactId>httpcore</artifactId>
      <version>4.4.9</version>
      <scope>provided</scope>
    </dependency>
`

The whole build environment is also very undeterministic. Sometimes, I have to delete openhab from the m2 repository to see an effect. Sometimes I can't add it anymore in the bnd.run. 
I had all dependencies in the pom.xml a long time ago, this didn't work. Then we said that the osgiify library might work better and the dependencies were removed from pom.xml. Now I should add them again? I also got completly confused if I should use a different scope for testing or not etc.

After 70 comments and no step further it's really time to make a decision. I came to work on features for openhab, but I learned that the new build environment is not very friendly to external libraries (the previous one worked like a charm).

If there is nobody else who has any idea how to proceed here, I'll close this PR in the next days (without any commits).

J-N-K commented:
I can try to resolve these issues maybe this weekend. Currently busy in my garden.

radokristof commented:
@J-N-K were you able to solve the dependency issues?

kaikreuzer commented:
@J-N-K Do you have this still on your list?

J-N-K commented:
Oops. Forgot about that. I'll have a look. But I guess that all these 3rd-party dependency issues (or at least the vast majority) will vanish when we merge something like #6004 or #6005. There is no need for resolution then, no need for wrapping or something else, because all that is needed is already included in the bundle (except some cases where we use shared dependencies, but this could be mocked during development by changing the scope from provided to compile).

kaikreuzer commented:
@J-N-K I am not so sure that embedding jars would apply for this binding. Those dependencies are "standard" libs like httpcomponents and jersey. Those are readily available as OSGi bundles and possibly be used by other bindings as well. And the only specific lib (telegrambot) is available in the osgiify repo (as you added it there already).

J-N-K commented:
I‘m lost. I can get the karaf feature to resolve easily. I added all the dependencies to the pom and in theory the bnd resolver should be able to resolve that. However, it hangs forever. It took me some time to find out that the dependency to org.json/json is the culprit here. As soon as I add that to either the binding-pom or the app-pom the resolver will not work. I guess this is due to the strange version numbering, I‘ll check that this weekend but I‘m out of ideas how this might work out in a build latery

bjoernbrings commented:
So the binding itself works but the resolver in eclipse doesn't?

kaikreuzer commented:
@J-N-K I am also not quite clear on your last comment - is the build ok, but not the IDE? What is left to do here (besides rebasing @ZzetT)?

J-N-K commented:
I can‘t remember. I‘ll check again this evening. I guess we can get it to resolve now that we can embed problematic dependencies.

J-N-K commented:
Finally I embedded everything which unfortunately results in a 10 MB bundle. Maybe we can strip that a little bit. The good thing is: it runs in my setup and sends messages to my telegram account.

There reason why this is so difficult is that the telegram-library has a lot of dependencies (e.g. jersey and Jackson) which are incompatible with the version we use in our features (our jersey-feature depends on Jackson 2.4, our Jackson feature on 2.9) but Jackson with 2.9 is required by the telegram lib). We should cleanup our core-Features, this will likely get us in trouble in other places.

I‘ll push my version later today, needs a little cleanup.

ZzetT commented:
What's your opinion how to proceed here? Shall I have a look again at one of those other telegram libraries like [java-telegram-bot-api](https://github.com/pengrad/java-telegram-bot-api) which seem to have much less dependencies? Or do we go now with the 10MB bundle which conflicts with other versions of the OH libaries like jackson? Or do we say that we are not bothered about receiving telegram messages and go back to the OH1 action and convert that to OH2?
For the first point (the other library) I might be able to have a look at the weekend.

J-N-K commented:
It's all here since 09/29/19: https://github.com/J-N-K/openhab2-addons/tree/ZzetT_with_lib. 

I'm not sure if other libraries are any better. I would not spend to much time on that. I'm ok with this solution, bit maybe someone else should express his opinion.

ZzetT commented:
1. Thanks for your effort J-N-K that you created a bundle which can be used in the osgi environment.

2. I set up a fresh Eclipse environment for OH and used your branch for the addons. Without the binding I'm able to resolve everything without errors. But as soon as I add it to the pom.xml in the demo app, I get the following resolution error:

_[ERROR] Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=org.openhab.core.model.rule.runtime)
          ⇒ [org.openhab.core.model.rule.runtime version=2.5.0.201910130313]
              ⇒ osgi.wiring.package: (&(osgi.wiring.package=com.google.inject)(&(version>=1.3.0)(!(version>=2.0.0))))
                  ⇒ [com.google.inject version=4.2.2]
                      ⇒ osgi.wiring.package: (&(osgi.wiring.package=com.google.common.base)(&(version>=25.1.0)(!(version>=26.0.0))))_

Strange because this time it doesn't seem to come from the bundle itself. But why is it working fine  without it?

3. Instead, I tried to setup a Visual Studio Code environment and downloaded the OH2.5.0 M3 version and put the build jar in the addons folder. Indeed, I was able to configure it and I could also send a message. However, as soon as I restarted OH, the whole paperui wasn't usable anymore ("Jersey is not ready yet" errors). I first thought it's a problem with my environment, but then I found out that the Telegram library throws an error as soon as it is started:

_[ERROR] [Telegram Bots Api                   ] - BOTSESSION
org.telegram.telegrambots.meta.exceptions.TelegramApiRequestException: Error getting updates
	at org.telegram.telegrambots.meta.api.methods.updates.GetUpdates.deserializeResponse(GetUpdates.java:119) ~[210:org.openhab.binding.telegram:2.5.0.201910111736]
	at org.telegram.telegrambots.updatesreceivers.DefaultBotSession$ReaderThread.getUpdatesFromServer(DefaultBotSession.java:257) ~[210:org.openhab.binding.telegram:2.5.0.201910111736]
	at org.telegram.telegrambots.updatesreceivers.DefaultBotSession$ReaderThread.run(DefaultBotSession.java:186) [210:org.openhab.binding.telegram:2.5.0.201910111736]_

I assume that this has something to do with the jersey error, because when I remove the jar, the error is gone and everything behaves normally. I thought that it might be related to a missing dependency or anything like that but then I saw that even my prototype bot which worked for OH2.4 throws the same error. I finally found an issue on github for that library about that exception: https://github.com/rubenlagus/TelegramBots/issues/677

Long story short:
We had already so many problems with that library, it seems to be quite instable and has many dependencies. I would really recommend to switch to another one. 

How much time would be left for the next release/milestone?

J-N-K commented:
This is a general problem with different jackson versions that is currently under investigation. This affects all bundles that use the openhab core jackson feature. I have a PR in openhab-core which might solve that.

kaikreuzer commented:
> I'm not sure if other libraries are any better. I would not spend too much time on that. I'm ok with this solution, bit maybe someone else should express his opinion.

I'd agree with @J-N-K here. I don't mind the bundle to be >10MB if it works. Refactoring the code to a new library could mean a complete rewrite of the binding, which can potentially take a lot of time.

Wrt Jackson: @J-N-K If the suitable Jackson lib is embedded, there is no issue with different versions being present in openHAB, correct? So if we embed it, is there still anything blocking wrt this PR?

ZzetT commented:
Yes, I agree "if it works" then the size of the bundle shouldn't matter so much. However, for me it seems that the library doesn't work anymore especially the part for the reception of messages. Maybe telegram changed something which is unexpected by that library. I have no other explanation at the moment.
Looking at the feedback in the telegram thread of the OH forum, I'm also the getting the feeling that some posts are not related to the binding itself but more to problems of the library. At least on my current OH2.4 instance, I also see that messages are dropped sporadically.

> could mean a complete rewrite of the binding, which can potentially take a lot of time.

I don't really fear to re-write the Java code. I know what's in there and it's more or less only about the TelegramActions class which is the interface between OH and the library. The thing that took most of the time was not the development but the build environment and the configuration of the dependencies. I just want to contribute something where everybody is happy with.

kaikreuzer commented:
Ok, if the current library does NOT work and you cannot even receive messages with it anymore, that's certainly a strong argument for refactoring it to use another lib.

Wrt your question about the timing: I'd roughly say that the binding should be done, reviewed and merged by mid of November to make it into the 2.5 release. Not sure if this is enough time for the endeavor...?

radokristof commented:
For me the last built jar is working and I can receive messages. Only the replies don't work.

radokristof commented:
I mean the last jar which was built and published on the forum by @ZzetT

ZzetT commented:
I finally switched to the other library from [pengrad](https://github.com/pengrad/java-telegram-bot-api). I have a much better feeling with this. Setting up the pom and features worked immediately. The code from that library looks a bit cleaner and simpler to me. The resulting size of that binding is less than 1MB! I posted it in the forum and would like to wait for some feedback before I merge it to the master branch.

kaikreuzer commented:
That sounds awesome, thanks! Could you maybe also rebase your branch so that the merge conflict is resolved and the jar is automatically build by our PR build?

clinique commented:
Why is lastMessageDate a String channel ?

ZzetT commented:
> Why is lastMessageDate a String channel ?

Thanks for the hint, I changed that now.

From my side I would be ready to merge it to the master unless you guys want to have another look. Since the last review a few things have changed (library replacement).

Hilbrand commented on bundles/org.openhab.binding.telegram/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,79 @@
+			</parameter>
+			<parameter name="parseMode" type="text" required="false">
+				<label>parseMode</label>
+				<description>Support for formatted messages, values: Markdown or HTML. Default: no formatting is used.</description>
```
Make this an options parameter? Something like this:
```suggestion
				<description>Support for formatted messages, values: Markdown or HTML. Default: no formatting is used.</description>
				<options>
					<option value="">No Formatting</option>
					<option value="HTML">HTML</option>
					<option value="Markdown">Markdown</option>
				</options>
```

ZzetT replied:
Thanks, good improvement! I changed it now.

ZzetT commented on bundles/org.openhab.binding.telegram/pom.xml:
```diff
@@ -0,0 +1,56 @@
+  <name>openHAB Add-ons :: Bundles :: Telegram Binding</name>
+
+  <properties>
+    <jackson.version>2.9.9</jackson.version>
```
Yes, they are not needed anymore (leftover from previous library). I removed them.

ZzetT commented on bundles/org.openhab.binding.telegram/src/main/feature/feature.xml:
```diff
@@ -0,0 +1,14 @@
+        <feature>openhab-runtime-base</feature>
+        <bundle start-level="80">mvn:org.openhab.addons.bundles/org.openhab.binding.telegram/${project.version}</bundle>
+        <bundle start-level="80">wrap:mvn:com.github.pengrad/java-telegram-bot-api/4.4.0</bundle>
+        <bundle start-level="80">mvn:com.google.code.gson/gson/2.8.5</bundle>
```
removed

ZzetT commented on bundles/org.openhab.binding.telegram/src/main/java/org/openhab/binding/telegram/bot/TelegramActions.java:
```diff
@@ -0,0 +1,400 @@
+            Integer messageId = localHandler.removeMessageId(chatId, replyId);
+            logger.debug("remove messageId {} for chatId {} and replyId {}", messageId, chatId, replyId);
+
+            // for (Long chatId : localHandler.getChatIds()) {
```
fixed
